### ISSUE TITLE
RequireThis false positive with shadowing try-with-resources variable

### ISSUE THREAD

**Issue by Kiena** [checkstyle/checkstyle#6845]
```
/var/tmp $ javac RequireThisAndTryWithResources.java

/var/tmp $ cat config.xml
<!DOCTYPE module PUBLIC '-//Puppy Crawl//DTD Check Configuration 1.3//EN' 'http://www.puppycrawl.com/dtds/configuration_1_3.dtd'>
<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="checkMethods" value="false"/>
            <property name="validateOnlyOverlapping" value="false"/>
        </module>
    </module>
</module>

/var/tmp $ cat RequireThisAndTryWithResources.java
public class RequireThisAndTryWithResources {
    private int value;

    public void printVariable() throws Exception {
        final AutoCloseable value = () -> System.out.println("Closing.");
        try (value) {
            System.out.println(value); // No violation.
        }
    }

    public void printResource() throws Exception {
        try (AutoCloseable value = () -> System.out.println("Closing.")) {
            System.out.println(value); // Violation???
        }
    }
}

/var/tmp $ java -Duser.language=en -Duser.country=US -jar checkstyle-8.22-all.jar -c config.xml RequireThisAndTryWithResources.java
Starting audit...
[ERROR] P:\java\workspaces\bugs\checkstyle\RequireThis\RequireThisAndTryWithResources.java:13:32: Reference to instance variable 'value' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

I expect no violation to be found in the `printResource` method. The `printVariable` method shows that the expected happens when the variable is declared before the try-with-resources statement (tested with Java 11).

**Comment by romani** [comment]
thanks a lot for good issue report.

Whole logic is located at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java
it is very easy to start debugging - https://checkstyle.org/beginning_development.html

if you have a time, please send us PR.

**Comment by romani** [comment]
Referenced code is almost complete fix, but still not up to required level.
Everyone is welcome to continue and finish issue.

**Comment by wltan** [comment]
I am on it. There seems to be some problems with pitest in #6860, will look into that to see what's going on.

**Comment by wltan** [comment]
I tried mirroring the changes in #6860, but it seems to be failing our current set of UTs. I'm guessing something else has been changed since then and this fix is no longer viable.

Going to unassign myself from this for now, anyone is free to pick up this issue.

**Comment by ZaheerAhmadDev** [comment]
@romani,
````/Desktop/falsepositive/src (main)
$ cat config.xml
<!DOCTYPE module PUBLIC '-//Puppy Crawl//DTD Check Configuration 1.3//EN' 'http://www.puppycrawl.com/dtds/configuration_1_3.dtd'>
<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="checkMethods" value="false"/>
            <property name="validateOnlyOverlapping" value="false"/>
        </module>
    </module>
</module>

/Desktop/falsepositive/src (main)
$ cat RequireThisAndTryWithResources.java
public class RequireThisAndTryWithResources {
    private int value;

    public void printVariable() throws Exception {
        final AutoCloseable value = () -> System.out.println("Closing.");
        try (value) {
            System.out.println(value); // No violation.
        }
    }

    public void printResource() throws Exception {
        try (AutoCloseable value = () -> System.out.println("Closing.")) {
            System.out.println(value); // Violation???
        }
    }
}

 /Desktop/falsepositive/src (main)
$  java -Duser.language=en -Duser.country=US -jar checkstyle-13.4.0-all.jar -c config.xml RequireThisAndTryWithResources.java
Starting audit...
Audit done.
````
The Issue has been solved. Checked it on new version.
I think it can be closed.?
